src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPrunerPass.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\DependencyInjection\Compiler;
+
+use Symfony\Component\Cache\PruneableInterface;
+use Symfony\Component\DependencyInjection\Argument\IteratorArgument;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\Exception\InvalidArgumentException;
+use Symfony\Component\DependencyInjection\Reference;
+
+class CachePoolPrunerPass implements CompilerPassInterface
+{
+    private $cacheCommandServiceId;
+    private $cachePoolTag;
+
+    public function __construct($cacheCommandServiceId = 'cache.command.pool_pruner', $cachePoolTag = 'cache.pool')
+    {
+        $this->cacheCommandServiceId = $cacheCommandServiceId;
+        $this->cachePoolTag = $cachePoolTag;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function process(ContainerBuilder $container)
+    {
+        if ($container->hasDefinition($this->cacheCommandServiceId)) {
+            $services = $this->getCachePoolServiceIds($container);
+            $iterator = new IteratorArgument(array_combine($services, array_map(function ($id) {
+                return new Reference($id);
+            }, $services)));
+
+            $container->getDefinition($this->cacheCommandServiceId)->replaceArgument(0, $iterator);
+        }
+    }
+
+    /**
+     * @param ContainerBuilder $container
+     *
+     * @return string[]
+     */
+    private function getCachePoolServiceIds(ContainerBuilder $container)
+    {
+        $services = array();
+
+        foreach ($container->findTaggedServiceIds($this->cachePoolTag) as $id => $tags) {
+            if ($container->getDefinition($id)->isAbstract()) {
+                continue;
+            }
+
+            if (!$this->resolveServiceReflectionClass($container, $id)->implementsInterface(PruneableInterface::class)) {
+                continue;
+            }
+
+            $services[] = $id;
```

Comment:
since this is internal, it should directly build the right structure: an id-to-ref map (and the array_combine above be remove)
also, I don't feel the need do keep this function, now that the caller is almost empty now. In should be merged with "process"